consensus/hotstuff/committees/consensus_committee.go

```diff
@@ -15,38 +13,43 @@ import (
 	"github.com/onflow/flow-go/module/irrecoverable"
 	"github.com/onflow/flow-go/state/protocol"
 	"github.com/onflow/flow-go/state/protocol/events"
-	"github.com/onflow/flow-go/state/protocol/prg"
 )
 
-// staticEpochInfo contains leader selection and the initial committee for one epoch.
-// This data structure must not be mutated after construction.
-type staticEpochInfo struct {
-	firstView            uint64                  // first view of the epoch (inclusive)
-	finalView            uint64                  // final view of the epoch (inclusive)
-	randomSource         []byte                  // random source of epoch
-	leaders              *leader.LeaderSelection // pre-computed leader selection for the epoch
-	initialCommittee     flow.IdentitySkeletonList
-	initialCommitteeMap  map[flow.Identifier]*flow.IdentitySkeleton
-	weightThresholdForQC uint64 // computed based on initial committee weights
-	weightThresholdForTO uint64 // computed based on initial committee weights
-	dkg                  hotstuff.DKG
+// epochInfo caches data about one epoch that is pertinent to the consensus committee.
+type epochInfo struct {
+	*leader.LeaderSelection // pre-computed leader selection for the epoch
+	initialCommittee        flow.IdentitySkeletonList
+	initialCommitteeMap     map[flow.Identifier]*flow.IdentitySkeleton
+	weightThresholdForQC    uint64 // computed based on initial committee weights
+	weightThresholdForTO    uint64 // computed based on initial committee weights
+	dkg                     hotstuff.DKG
 }
 
-// newStaticEpochInfo returns the static epoch information from the epoch.
-// This can be cached and used for all by-view queries for this epoch.
-func newStaticEpochInfo(epoch protocol.Epoch) (*staticEpochInfo, error) {
-	firstView, err := epoch.FirstView()
+// recomputeLeaderSelectionForExtendedViewRange re-computes the LeaderSelection field
+// for the input epoch's entire view range, including extensions.
+// This should be called each time an extension is added to an epoch.
+// No errors are expected during normal operation.
+func (e *epochInfo) recomputeLeaderSelectionForExtendedViewRange(epoch protocol.Epoch) error {
+	extendedFinalView, err := epoch.FinalView()
 	if err != nil {
-		return nil, fmt.Errorf("could not get first view: %w", err)
+		return fmt.Errorf("could not get final view for extended epoch: %w", err)
 	}
-	finalView, err := epoch.FinalView()
-	if err != nil {
-		return nil, fmt.Errorf("could not get final view: %w", err)
+	// sanity check: ensure view range for which leader selection is defined strictly increases
+	if extendedFinalView <= e.FinalView() {
+		return nil
 	}
-	randomSource, err := epoch.RandomSource()
+
+	leaderSelection, err := leader.SelectionForConsensus(epoch)
```

Comment:
Since we have already removed the convention that `epochInfo` is immutable, `epochInfo` could also hold a reference to the CSPRNG that is used for the leader selection. Thereby, we wouldn't need to re-compute 750k leaders each time we add an extension. I think the algorithmic overhead of this optimization is minimal:
* initialize the PRNG (currently [this code](https://github.com/onflow/flow-go/blob/d724014621cf4e9a6a54e4e6122e70a517cc1d74/consensus/hotstuff/committees/leader/consensus.go#L29) in the constructor `newEpochInfo`
* Then, I think `SelectionForConsensus` should become a private method of `epochInfo` because for correctness, and consistency the PRNG must be initialized with epoch-specific parameters. But in the current code base, the method  `SelectionForConsensus` is anyway only called by the `epochInfo`. So in terms of business logic, it would be a negligible refactoring. Nevertheless, there are some changes in the test necessary.

Reply:
I agree that we could store the PRNG for subsequent calls. But `ComputeLeaderSelection`, which accepts the PRNG, is set-up to return a new `LeaderSelection` for each call, so there would be additional changes needed. What I had in mind was changing the way view ranges are stored in the Consensus Committee to accommodate multiple `LeaderSelection`'s per epoch, which seemed non-trivial. 

Given the concrete cost and frequency of computing leader selection I didn't think it was worthwhile to explore further. For reference, computing leader selection for one epoch on Mainnet takes about 20-30ms, including scheduled view rate increases.

However, after thinking about your comment, I'm wondering if adding a function to merge 2 consecutive `LeaderSelection`s would be sufficient. I think this would be simpler than tracking multiple `LeaderSelection` instances per epoch. We would instantiate a new leader selection only for the view range of the extension, then merge it with the current leader selection for the epoch.

Reply:
I took a closer look and came to the same conclusion as you: based on the engineering cost compared to the frequency of needing to extend leader selections, its not worth it. 
I think there might be value in consolidating the logic for. the leader selections, purely for readability and encapsulation (reducing surface for possible bugs). Though, I think we should leave that for a later PR (if we event want to do it).